src/mpi/init/initthread.c

```diff
@@ -76,120 +76,140 @@ int MPI_Init_thread(int *argc, char ***argv, int required, int *provided)
    how MPICH was configured. */
 extern const char MPII_Version_device[];
 
-int MPIR_Init_thread(int *argc, char ***argv, int required, int *provided)
+int MPIR_Init_thread(int *argc, char ***argv, int user_required, int *provided)
 {
     int mpi_errno = MPI_SUCCESS;
+    int required = user_required;
+    int err;
 
-    MPII_init_thread_and_enter_cs();
-    /* ---- Entered CS ------------------------------------------------ */
+    /**********************************************************************/
+    /* Section 1: base components that other components rely on.
+     * These need to be intialized first.  They have strong
+     * dependencies between each other, which makes it a pain to
+     * maintain them.  Hopefully, the number of such components is
+     * small. */
+    /**********************************************************************/
 
     MPL_wtime_init();
-    MPII_pre_init_dbg_logging(argc, argv);
-
-    mpi_errno = MPIR_T_env_init();
-    MPIR_ERR_CHECK(mpi_errno);
 
-    MPII_init_request();
+    MPID_Thread_init(&err);
+    MPIR_Assert(err == 0);
 
-    /* Certain features may potentially change required thread-level */
-    mpi_errno = MPII_init_global(&required);
-    MPIR_ERR_CHECK(mpi_errno);  /* out-of-mem */
 
-    /* Device layer parse command line, decide thread level, and preset configurations. */
-    mpi_errno = MPID_Pre_init(argc, argv, required, &MPIR_ThreadInfo.thread_provided);
-    MPIR_ERR_CHECK(mpi_errno);
-    if (MPIR_ThreadInfo.thread_provided > MPICH_THREAD_LEVEL) {
-        MPIR_ThreadInfo.thread_provided = MPICH_THREAD_LEVEL;
-    }
+    /**********************************************************************/
+    /* Section 2: contains initialization of local components (no
+     * dependency on other processes) that do not need any information
+     * from the device.  These components are independent of each
+     * other and can be initialized in any order. */
+    /**********************************************************************/
 
-    MPL_set_threaded(MPIR_ThreadInfo.thread_provided == MPI_THREAD_MULTIPLE);
+    MPIR_Err_init();
+    MPII_thread_mutex_create();
```

Comment:
When other init step jumps to fail, we need destroy the mutexes, so the creation of them has to happen before any possible error jump. So  mutex creation and `MPIR_Err_init` are a dependencies for other steps, shouldn't they belong in section 1?

Reply:
I don't want to add too much stuff inside section 1.  My preference is to say that all errors in `MPI_INIT` are fatal and we'll abort the application if that happens.

Reply:
When you explicitly comment that `These components are independent of each other and can be initialized in any order ` then you need stick to that assertion, right? Does `goto fn_fail` depend on `MPIR_Err_init()` and `MPII_thread_mutex_create()`?